src/bin/ion/commands/jq.rs

```diff
@@ -173,11 +206,141 @@ impl PartialOrd for JaqElement {
+        Decimal::new(data, -exponent)
+    }
+}
+
+trait FloatMath {
+    fn to_f64(self) -> Option<f64>;
+}
+
+impl FloatMath for f64 {
+    fn to_f64(self) -> Option<f64> {
+        Some(self)
+    }
+}
+
+impl FloatMath for Int {
+    fn to_f64(self) -> Option<f64> {
+        self.as_i128().and_then(|data| {
+            let float = data as f64;
+            (float as i128 == data).then_some(float)
+        })
+    }
+}
+
+impl FloatMath for Decimal {
+    fn to_f64(self) -> Option<f64> {
+        self.to_big_decimal().to_f64()
+    }
+}
+
+impl FloatMath for Value {
+    fn to_f64(self) -> Option<f64> {
+        match self {
+            Value::Int(i) => i.to_f64(),
+            Value::Decimal(d) => d.to_f64(),
+            _ => None,
+        }
+    }
+}
+
 impl Add for JaqElement {
     type Output = ValR<Self>;
 
     fn add(self, _rhs: Self) -> Self::Output {
-        todo!()
+        let (lhv, rhv) = (self.into_value(), _rhs.into_value());
+
+        use Value::*;
+
+        let elt: Element = match (lhv, rhv) {
+            (List(a), List(b)) => ion_rs::List::from_iter(a.into_iter().chain(b)).into(),
+            (SExp(a), SExp(b)) => ion_rs::SExp::from_iter(a.into_iter().chain(b)).into(),
+            (String(a), String(b)) => format!("{}{}", a.text(), b.text()).into(),
+            //TODO: Recursively remove duplicate fields, first field position but b and last field wins
+            (Struct(a), Struct(b)) => a.clone_builder().with_fields(b.fields()).build().into(),
+
+            // Number types, only lossless operations
+            (Int(a), Int(b)) => (a + b).into(),
+            (Float(a), Float(b)) => (a + b).into(),
+            (Decimal(a), Decimal(b)) => a.add(b).into(),
+            (Decimal(a), Int(b)) | (Int(b), Decimal(a)) => a.add(b).into(),
+
+            // jq treats JSON's untyped null as an additive identity, e.g. 0 / "" / [] / {}
+            (Null(IonType::Null), a) | (a, Null(IonType::Null)) => a.into(),
+
+            // Typed nulls we must handle differently, we can only add similar types
+            (Null(a), Null(b)) if a == b => Null(a).into(),
+            (Null(a), b) | (b, Null(a)) if a == b.ion_type() => b.into(),
+
+            // Only try potentially lossy Float conversions when we've run out of the other options
+            (b, Float(a)) | (Float(a), b)
+                if matches!(b.ion_type(), IonType::Int | IonType::Decimal) =>
+            {
+                let Some(f) = b.clone().to_f64() else {
+                    return jaq_error(format!("{b:?} cannot be an f64"));
+                };
+                (f + a).into()
+            }
+
+            //TODO: Better error messaging (display for JaqElement?)
+            // Note this includes timestamps
+            (a, b) => {
+                let atype = a.ion_type();
+                let btype = b.ion_type();
+                return jaq_error(format!("{atype} ({a}) and {btype} ({b}) cannot be added"));
+            }
```

Comment:
Is there a better way to do this? I want a consistent way to turn a `Value` into essentially `format!("{self.ion_type()} ({self})")`